components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.scheduler
+
+import android.content.Context
+import android.content.SharedPreferences
+import mozilla.components.support.base.log.logger.Logger
+import java.util.concurrent.TimeUnit
+
+/**
+ * MetricsPingScheduler facilitates sending and scheduling of metrics pings
+ */
+internal class MetricsPingScheduler(val applicationContext: Context) {
+    private val logger = Logger("glean/MetricsPingScheduler")
+    internal val sharedPreferences: SharedPreferences by lazy { getMetricsPrefs() }
+
+    companion object {
+        const val LAST_METRICS_PING_TIMESTAMP_KEY = "last_metrics_ping_timestamp"
+        const val STORE_NAME = "metrics"
+        const val PING_INTERVAL_HOURS = 24L
+    }
+
+    /**
+     * Check to see if a metrics ping can be sent.
+     *
+     * This will compare the current system time to the last ping time stored in [SharedPreferences]
+     * to determine whether [PING_INTERVAL_HOURS] has elapsed.
+     *
+     * @return true if the time interval has elapsed or if this is the first attempt to send a
+     *         metrics ping, otherwise false.
+     */
+    fun canSendPing(): Boolean {
+        var lastPingTime = 0L
+
+        try {
+            lastPingTime = sharedPreferences.getLong(LAST_METRICS_PING_TIMESTAMP_KEY, 0L)
+        } catch (e: ClassCastException) {
+            // Intentionally doing nothing here, if LAST_METRICS_PING_TIMESTAMP_KEY's value is
+            // not a Long, then we will let lastPingTime remain 0L for the purposes of comparison
+            logger.error("MetricsPingScheduler last ping time stored was not a Long value")
+        }
+
+        val currentTime = System.currentTimeMillis()
+
+        if (lastPingTime == 0L ||
+            currentTime - lastPingTime >= TimeUnit.HOURS.toMillis(PING_INTERVAL_HOURS)) {
+            return true
+        }
+
+        return false
+    }
+
+    /**
+     * Update the persisted timestamp when the metrics ping is sent.
+     *
+     * This is called after sending a metrics ping in order to timestamp when the last ping was
+     * sent in order to maintain the proper interval between pings. See [canSendPing].
+     *
+     * @param timestamp The timestamp to store, which defaults to [System.currentTimeMillis].
+     * @return Returns the time that was stored in [SharedPreferences] for comparision.
+     */
+    fun updateSentTimestamp(timestamp: Long = System.currentTimeMillis()): Long {
```

Comment:
Is this return value only for tests? I'm not sure we need this though. See my comments below.

Reply:
Yes this return value is only for tests, it was a simple way to test the default and have a way to compare it.